File: stepless.py
from __future__ import print_function

from collections import defaultdict
import itertools
import sys

import clingo

def make_sym_map(iteratoms):
    result = defaultdict(list)
    for s in iteratoms:
        result[s.name].append(s.arguments)
    return result

def toposort(vertices, edges):
    visited = set()
    def sortFrom(vertex):
        if vertex in visited:
            return
        visited.add(vertex)
        for e in edges[vertex]:
            for x in sortFrom(e):
                yield x
        yield vertex
    result = list(itertools.chain(*(map(sortFrom,vertices))))
    result.reverse()
    return result

def print_action_plan(symbols,file):
    vertices = set(v for [v] in symbols['vertex'])
    eventToVertex = {k:v for [k,v] in symbols['inVertex']}
    edges = defaultdict(list)
    for [vfrom,vto] in symbols['edge']:
        edges[eventToVertex[vfrom]].append(eventToVertex[vto])
    verts = toposort(vertices, edges)
    actions = [v for v in verts if v.name == 'actOcc']
    for a in actions:
        print(a.arguments[0],file=file)

def print_suffix_plan(symbols,file):
    actions = {v for [v] in symbols['happens']}
    vertices = actions.union(set(v for [v] in symbols['holds']))
    edges = defaultdict(list)
    for [vfrom,vto] in symbols['causes'] + symbols['permits']:
        edges[vfrom].append(vto)
    verts = toposort(vertices,edges)
    actions = [v for v in verts if v in actions]
    for v in verts:
        if v in actions:
            print('suffix',v,file=file)

def print_model_plan(m, file):
    symbols = make_sym_map(m.symbols(shown=True))
    print_action_plan(symbols,file)
    suffix_symbols = make_sym_map(x for [x] in symbols['suffix'])
    print_suffix_plan(suffix_symbols,file)
    print('cost:',m.cost[0])

def new_occs(m):
    for s in m.symbols(atoms=True):
        if s.name == 'saturated':
            arg = s.arguments[0]
            if arg.name == 'action':
                r = ('actOcc', arg.arguments[0])
                yield r
            elif arg.name == 'fluent':
                r = ('fluentOcc', arg.arguments[0])
                yield r

def rule_of(k0,v):
    (t,k) = k0
    return str(clingo.Function('is', [clingo.Function(t,[k,v])])) + '.'

def main(*infiles):
    occs = defaultdict(lambda: 1)
    while True:
        c = clingo.Control(['--opt-strategy=usc'])
        for f in infiles:
            c.load(f)
        c.load("asp/stepless.asp")
        parts = [(t,[k,i]) for ((t,k),v) in occs.iteritems() for i in range(2,v+1)]
        c.ground([("base",[])] + parts)
        with c.solve(yield_=True) as handle:
            for m in handle:
                pass
            res = handle.get()
            if res.interrupted:
                break
            elif res.satisfiable:
                if not m.contains(clingo.Function('useSuffix',[])):
                    print_model_plan(m, sys.stdout)
                    break
                print_model_plan(m, sys.stderr)
                print(file=sys.stderr)
                print('Adding:',file=sys.stderr)
                for k in new_occs(m):
                    v = occs[k] + 1
                    occs[k] = v
                    print(rule_of(k,v),file=sys.stderr)
                print(file=sys.stderr)
                print(file=sys.stderr)
            else:
                print('no solution')
                break


if __name__ == '__main__':
    main(*sys.argv[1:])

File: mcp/src/MCP/Solver/Stepless.hs
module MCP.Solver.Stepless
  ( solve
  )
where

import           Clingo.Control                 ( Clingo
                                                , ClingoSetting(ClingoSetting)
                                                )
import qualified Clingo.Control                as Clingo
import           Clingo.Model
import qualified Clingo.Symbol                 as Clingo
import           Control.Monad                  ( forM )
import           Data.Map                       ( Map )
import qualified Data.Map                      as Map
import           Data.Maybe
import           Data.Text                      ( Text )
import qualified Data.Text.Lazy                as LText
import           Debug.Trace
import           GHC.Stack                      ( HasCallStack )
import           System.FilePath                ( (</>) )

import           IData.Graph
import qualified IData.IMap                    as IMap
import           MCP.ASPInput
import           MCP.Config
import           MCP.Solver.Output              ( actionStatement )
import           MCP.Util
import           PDDLParser.Translate.Shared    ( ReverseMap )

solve :: SteplessOptions -> ASPInput -> ReverseMap -> IO LText.Text
solve SteplessOptions { useUSC } inp rev = do
  aspPath <- getASPPath
  let
    go occs = do
      result <-
        Clingo.withClingo
            (ClingoSetting [ "--opt-strategy=usc" | useUSC ] Nothing 0)
          $ do
              parts <- partsFrom occs
              setup
                (inp <> ASPInput [ASPFile $ aspPath </> "stepless.asp"] Nothing)
                parts
              Clingo.ground parts Nothing
              solveAwaitLastModel Nothing $ \m -> do
                useSuffix    <- Clingo.createFunction "useSuffix" [] True
                hasUseSuffix <- m `contains` useSuffix
                syms         <- modelSymbols m selectNone { selectShown = True }
                let pureSyms = map Clingo.toPureSymbol syms
                cost <- Cost <$> costVector m
                return (pureSyms, cost, hasUseSuffix)
      case result of
        Nothing                     -> return Nothing
        Just (syms, cost, suffUsed) -> do
          traceShowM cost
          if suffUsed
            then do
              traceM $ LText.unpack (buildSolution rev (Just syms))
              let sat = saturatedOccs syms
              go (incrEach sat occs)
            else return (Just syms)
  buildSolution rev <$> go Map.empty

buildSolution :: ReverseMap -> Maybe [Clingo.PureSymbol] -> LText.Text
buildSolution rev = \case
  Nothing -> "No Solution"
  Just syms ->
    let LabeledGraph g labels = constructGraph syms
        actionSym             = \case
          Normal (Clingo.PureFunction "actOcc" [act, _] True) ->
            Just (Normal act)
          Suffix (Clingo.PureFunction "happens" [act] True) ->
            Just (Suffix act)
          _ -> Nothing
        sortedVerts = mapMaybe (actionSym . (labels IMap.!)) $ topologicSort g
    in  LText.unlines
          $ mapMaybe (fmap LText.fromStrict . actionStatement' rev) sortedVerts

actionStatement' :: ReverseMap -> PlanVertex -> Maybe Text
actionStatement' rev = \case
  Normal x -> actionStatement rev x
  Suffix x -> ("suffix " <>) <$> actionStatement rev x

type OccurrenceMap = Map Clingo.PureSymbol Integer

partsFrom :: OccurrenceMap -> Clingo s [Clingo.Part s]
partsFrom m = do
  let allOccs = [ (k, i) | (k, v) <- Map.toList m, i <- [2 .. v] ]
  forM allOccs $ \(o, i) -> case o of
    Clingo.PureFunction "action" [act] True -> Clingo.Part "actOcc"
      <$> sequence [Clingo.unpureSymbol act, Clingo.createNumber i]
    Clingo.PureFunction "fluent" [flu] True -> Clingo.Part "fluentOcc"
      <$> sequence [Clingo.unpureSymbol flu, Clingo.createNumber i]
    _ -> error $ "Unknown saturation " ++ show o

saturatedOccs :: [Clingo.PureSymbol] -> [Clingo.PureSymbol]
saturatedOccs = mapMaybe justIfSaturated
 where
  justIfSaturated = \case
    Clingo.PureFunction "saturated" [arg] True -> Just arg
    _ -> Nothing

-- Counter defaults to 1
incrEach :: Ord a => [a] -> Map a Integer -> Map a Integer
incrEach = Map.unionWith (\_ oldVal -> oldVal + 1) . Map.fromList . map (, 2)

data PlanVertex = Normal Clingo.PureSymbol | Suffix Clingo.PureSymbol
  deriving (Eq, Ord, Show)

constructGraph :: [Clingo.PureSymbol] -> LabeledGraph PlanVertex
constructGraph syms = mkGraph verts edges
 where
  verts   = mapMaybe getVert syms
  getVert = \case
    Clingo.PureFunction "vertex" [vert] True -> Just (Normal vert)
    Clingo.PureFunction "suffix" [vert@(Clingo.PureFunction "holds" _ True)] True
      -> Just (Suffix vert)
    Clingo.PureFunction "suffix" [vert@(Clingo.PureFunction "happens" _ True)] True
      -> Just (Suffix vert)
    _ -> Nothing
  inVerts   = Map.fromList $ mapMaybe getInVert syms
  getInVert = \case
    Clingo.PureFunction "inVertex" [x, y] True -> Just (Normal x, Normal y)
    _ -> Nothing
  edges   = mapMaybe getEdge syms
  getEdge = \case
    Clingo.PureFunction "edge" [x, y] True ->
      Just (inVerts ! Normal x, inVerts ! Normal y)
    Clingo.PureFunction "suffix" [Clingo.PureFunction "causes" [x, y] True] True
      -> Just
        ( Suffix (Clingo.PureFunction "happens" [x] True)
        , Suffix (Clingo.PureFunction "holds" [y] True)
        )
    Clingo.PureFunction "suffix" [Clingo.PureFunction "permits" [x, y] True] True
      -> Just
        ( Suffix (Clingo.PureFunction "holds" [x] True)
        , Suffix (Clingo.PureFunction "happens" [y] True)
        )
    Clingo.PureFunction "permits" [x, Clingo.PureFunction "subgoal" [y] True] True
      -> Just
        ( Normal (Clingo.PureFunction "holds" [x] True)
        , Suffix (Clingo.PureFunction "holds" [y] True)
        )
    _ -> Nothing

(!) :: HasCallStack => (Ord k, Show k) => Map k v -> k -> v
(!) m k = fromMaybe (error $ "Could not find key: " ++ show k) $ m Map.!? k
